sql-plugin/src/main/java/com/nvidia/spark/rapids/iceberg/spark/source/GpuMultiFileBatchReader.java

```diff
@@ -0,0 +1,252 @@
+  private final boolean useMultiThread;
+  private final FileFormat fileFormat;
+  private final int numThreads;
+  private final int maxNumFileProcessed;
+
+  private NameMapping nameMapping = null;
+  private boolean needNext = true;
+  private boolean isBatchPending;
+  // lazy variables
+  private FilePartitionReaderBase rapidsReader = null;
+
+  GpuMultiFileBatchReader(CombinedScanTask task, Table table, Schema expectedSchema,
+      boolean caseSensitive, Configuration conf, int maxBatchSizeRows, long maxBatchSizeBytes,
+      String parquetDebugDumpPrefix, int numThreads, int maxNumFileProcessed,
+      boolean useMultiThread, FileFormat fileFormat,
+      scala.collection.immutable.Map<String, GpuMetric> metrics) {
+    super(table, task);
+    this.expectedSchema = expectedSchema;
+    this.caseSensitive = caseSensitive;
+    this.conf = conf;
+    this.maxBatchSizeRows = maxBatchSizeRows;
+    this.maxBatchSizeBytes = maxBatchSizeBytes;
+    this.parquetDebugDumpPrefix = parquetDebugDumpPrefix;
+    this.useMultiThread = useMultiThread;
+    this.fileFormat = fileFormat;
+    this.metrics = metrics;
+    this.numThreads = numThreads;
+    this.maxNumFileProcessed = maxNumFileProcessed;
+    String nameMapping = table.properties().get(TableProperties.DEFAULT_NAME_MAPPING);
+    if (nameMapping != null) {
+      this.nameMapping = NameMappingParser.fromJson(nameMapping);
+    }
+    files = Maps.newLinkedHashMapWithExpectedSize(task.files().size());
+    task.files().forEach(fst -> this.files.putIfAbsent(fst.file().path().toString(), fst));
+  }
+
+  @Override
+  public ColumnarBatch get() {
+    if (rapidsReader == null) {
+      // Not initialized, return null to align with PerFile reader.
+      return null;
+    }
+    needNext = true;
+    isBatchPending = false;
+    // The same post-process with PerFile reader.
+    try (ColumnarBatch batch = rapidsReader.get()) {
+      // The Rapids reader should already set the current file.
+      String curFile = InputFileUtils.getCurInputFilePath();
+      // <idToConstants, updatedExpectedSchema>
+      Tuple2<Map<Integer, ?>, Schema> constsSchema = constsSchemaMap.get(curFile);
+      return GpuIcebergReader.addUpcastsIfNeeded(
+          GpuIcebergReader.addConstantColumns(batch, constsSchema._2(), constsSchema._1()),
+          constsSchema._2());
+    }
+  }
+
+  @Override
+  public boolean next() throws IOException {
+    ensureRapidsReader();
+    if (needNext) {
+      needNext = false;
+      isBatchPending = rapidsReader.next();
+    }
+    return isBatchPending;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (rapidsReader != null) rapidsReader.close();
+    super.close();
+  }
+
+  @Override
+  CloseableIterator<ColumnarBatch> open(FileScanTask task) {
+    // Stub for the required implementation.
+    // This method will never be called after overriding the "next" method.
+    throw new UnsupportedOperationException();
```

Comment:
May want to use IllegalStateException, since we don't ever expect to get here.

Reply:
Updated